dapps/src/lib.rs

```diff
@@ -101,31 +101,51 @@ impl<F> WebProxyTokens for F where F: Fn(String) -> Option<Origin> + Send + Sync
 }
 
 /// Current supported endpoints.
+#[derive(Default, Clone)]
 pub struct Endpoints {
 	endpoints: endpoint::Endpoints,
+	dapps_path: PathBuf,
+	local_endpoints: Vec<String>,
+	embeddable: Option<ParentFrameSettings>,
 }
 
 impl Endpoints {
 	/// Returns a current list of app endpoints.
 	pub fn list(&self) -> Vec<apps::App> {
-		self.endpoints.iter().filter_map(|(ref k, ref e)| {
+		self.endpoints.read().iter().filter_map(|(ref k, ref e)| {
 			e.info().map(|ref info| apps::App::from_info(k, info))
 		}).collect()
 	}
+	/// Check for any changes in the local dapps folder and update.
+	pub fn refresh_local_dapps(&mut self) {
```

Comment:
This method shouldn't be `&mut` the `local_endpoints` should be wrapped in `Arc<RwLock` instead (like `endpoints`).